blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,865 @@
+            None => Err(Error::Other("No actor state".to_string())),
+        }
+    }
+
+    fn messages_for_block(
+        &self,
+        h: &BlockHeader,
+    ) -> Result<(Vec<UnsignedMessage>, Vec<SignedMessage>), Error> {
+        chain::block_messages(self.cs.blockstore(), h).map_err(|err| Error::Other(err.to_string()))
+    }
+
+    fn messages_for_tipset(&self, h: &Tipset) -> Result<Vec<UnsignedMessage>, Error> {
+        let mut umsg: Vec<UnsignedMessage> = Vec::new();
+        let mut msg: Vec<SignedMessage> = Vec::new();
+        for bh in h.blocks().iter() {
+            let (mut bh_umsg_tmp, mut bh_msg_tmp) = self.messages_for_block(bh)?;
+            let bh_umsg = bh_umsg_tmp.as_mut();
+            let bh_msg = bh_msg_tmp.as_mut();
+            umsg.append(bh_umsg);
+            msg.append(bh_msg);
+        }
+        for msg in &msg {
+            umsg.push(msg.message().clone());
+        }
+        Ok(umsg)
+    }
+
+    fn load_tipset(&self, tsk: &TipsetKeys) -> Result<Tipset, Error> {
+        self.cs
+            .tipset_from_keys(tsk)
+            .map_err(|err| Error::Other(err.to_string()))
+    }
+}
+
+/// This is the main MessagePool struct TODO async safety
+pub struct MessagePool<T> {
+    local_addrs: Vec<Address>,
+    pending: HashMap<Address, MsgSet>,
+    pub(crate) cur_tipset: Tipset,
+    api: T,
+    pub min_gas_price: BigInt,
+    pub max_tx_pool_size: i64,
+    pub network_name: String,
+    bls_sig_cache: LruCache<Cid, Signature>,
+    sig_val_cache: LruCache<String, ()>,
+    local_msgs: HashMap<Vec<u8>, SignedMessage>,
+}
+
+impl<T> MessagePool<T>
+where
+    T: Provider,
+{
+    /// Create a new message pool
+    pub fn new(mut api: T, network_name: String) -> Result<MessagePool<T>, Error>
+    where
+        T: Provider,
+    {
+        let tipset = api.subscribe_head_changes();
+        // LruCache sizes have been taken from the lotus implementation
+        let bls_sig_cache = LruCache::new(40000);
+        let sig_val_cache = LruCache::new(32000);
+        let mut mp = MessagePool {
+            local_addrs: Vec::new(),
+            pending: HashMap::new(),
+            cur_tipset: tipset,
+            api,
+            min_gas_price: ToBigInt::to_bigint(&0).unwrap(),
+            max_tx_pool_size: 5000,
+            network_name,
+            bls_sig_cache,
+            sig_val_cache,
+            local_msgs: HashMap::new(),
+        };
+        mp.load_local()?;
+        Ok(mp)
+    }
+
+    /// Add a signed message to local_addrs and local_msgs
+    fn add_local(&mut self, m: &SignedMessage, msgb: Vec<u8>) -> Result<(), Error> {
+        self.local_addrs.push(*m.from());
+        self.local_msgs.insert(msgb, m.clone());
+        Ok(())
+    }
+
+    /// Push a signed message to the MessagePool
+    pub fn push(&mut self, msg: &SignedMessage) -> Result<Cid, Error> {
+        let msg_serial = msg
+            .marshal_cbor()
+            .map_err(|err| Error::Other(err.to_string()))?;
```

Comment:
Reading Lotus, it seems they return an Undefined CID here and use it as such. This might need to return an Option to check that CID. Will defer to @austinabell though, I didn't realize `cid.Undef` was acceptable.

Reply:
No don't think they do, should be fine as is. Can you show where they use the undef cid? They don't accept or allow Cids to be used in anything within the protocol, they just use for returns like this as it's empty data to return with the error (go sucks)